src/sync/generic_network_transport.hpp

```diff
@@ -0,0 +1,354 @@
+    }
+
+    const std::string message() const
+    {
+        return this->what();
+    }
+
+    int code() const
+    {
+        return m_code;
+    }
+
+    AppError()
+    : std::runtime_error("AppError")
+    , type(Type::Unknown)
+    , m_code(-1)
+    {
+    };
+    AppError(std::string msg, int code, Type classification = Type::Unknown)
+    : std::runtime_error(msg),
+    type(classification),
+    m_code(code)
+    {
+    }
+
+    const Type type;
+    
+private:
+    const int m_code;
+};
+
+enum class JSONErrorCode {
+    bad_token = 1,
+    malformed_json = 2,
+    missing_json_key = 3,
+
+    none = 0
+};
+
+struct JSONError : public AppError
+{
+
+
+    JSONError(JSONErrorCode c, std::string msg)
+    : AppError(msg, static_cast<int>(c), AppError::Type::JSON)
+    , code(c)
+    {
+    }
+
+    const JSONErrorCode code;
+};
+
+#define HAS_JSON_KEY_OR_THROW(JSON, KEY, RET_TYPE) \
+JSON.find(KEY) != JSON.end() ? JSON[KEY].get<RET_TYPE>() : \
+throw app::error::JSONError(app::error::JSONErrorCode::missing_json_key, KEY)
+
+enum class ServiceErrorCode {
+    missing_auth_req = 1,
+    /// Invalid session, expired, no associated user, or app domain mismatch
+    invalid_session = 2,
+    user_app_domain_mismatch = 3,
+    domain_not_allowed = 4,
+    read_size_limit_exceeded = 5,
+    invalid_parameter = 6,
+    missing_parameter = 7,
+    twilio_error = 8,
+    gcm_error = 9,
+    http_error = 10,
+    aws_error = 11,
+    mongodb_error = 12,
+    arguments_not_allowed = 13,
+    function_execution_error = 14,
+    no_matching_rule_found = 15,
+    internal_server_error = 16,
+    auth_provider_not_found = 17,
+    auth_provider_already_exists = 18,
+    service_not_found = 19,
+    service_type_not_found = 20,
+    service_already_exists = 21,
+    service_command_not_found = 22,
+    value_not_found = 23,
+    value_already_exists = 24,
+    value_duplicate_name = 25,
+    function_not_found = 26,
+    function_already_exists = 27,
+    function_duplicate_name = 28,
+    function_syntax_error = 29,
+    function_invalid = 30,
+    incoming_webhook_not_found = 31,
+    incoming_webhook_already_exists = 32,
+    incoming_webhook_duplicate_name = 33,
+    rule_not_found = 34,
+    api_key_not_found = 35,
+    rule_already_exists = 36,
+    rule_duplicate_name = 37,
+    auth_provider_duplicate_name = 38,
+    restricted_host = 39,
+    api_key_already_exists = 40,
+    incoming_webhook_auth_failed = 41,
+    execution_time_limit_exceeded = 42,
+    not_callable = 43,
+    user_already_confirmed = 44,
+    user_not_found = 45,
+    user_disabled = 46,
+
+    unknown = -1,
+    none = 0
+};
+
+/// Struct allowing for generic error data.
+struct ServiceError : public AppError
+{
+    ServiceError(std::string raw_code, std::string message)
+    : AppError(message, static_cast<int>(error_code_for_string(raw_code)), AppError::Type::Service)
+    {
+    }
+    const std::string message() const { return std::runtime_error::what(); };
```

Comment:
I don't understand what this is trying to do, but I'm pretty sure it's not doing anything compared to just calling the function defined in the parent class.